File: llm_written/go_gemini/gemm/gemm_parallel.go
package llm_written

import (
	"fmt"
	"runtime"
	"sync"
)

// Matrix is a rectangular 2D slice of float64.
type Matrix [][]float64

// getSize returns (rows, cols) of a rectangular matrix.
func getSize(m Matrix) (int, int) {
	if len(m) == 0 {
		return 0, 0
	}
	return len(m), len(m[0])
}

// validateMatrix checks that the matrix is non-empty and rectangular.
func validateMatrix(m Matrix, name string) error {
	if m == nil {
		return fmt.Errorf("%s must be a non-empty list of rows", name)
	}
	if len(m) == 0 {
		return fmt.Errorf("%s must have at least one row", name)
	}
	if len(m[0]) == 0 {
		return fmt.Errorf("%s must have at least one column", name)
	}
	cols := len(m[0])
	for r, row := range m {
		if row == nil {
			return fmt.Errorf("%s row %d is nil", name, r)
		}
		if len(row) != cols {
			return fmt.Errorf("%s is ragged: row %d has %d cols, expected %d", name, r, len(row), cols)
		}
	}
	return nil
}

// generateMatrix returns an rows x cols zero matrix.
func generateMatrix(rows, cols int) Matrix {
	out := make(Matrix, rows)
	for i := range out {
		out[i] = make([]float64, cols)
	}
	return out
}

// partialMatmul computes the matrix product for a small block and accumulates the result.
func partialMatmul(A, B, C Matrix, alpha float64, m0, n0, kb int) {
	for iOff, Aik := range A {
		Ci := C[m0+iOff]
		for jOff, Bjk := range B {
			s := 0.0
			for k := 0; k < kb; k++ {
				s += Aik[k] * Bjk[k]
			}
			Ci[n0+jOff] += alpha * s
		}
	}
}

// transpose returns the transpose of a rectangular matrix.
func transpose(m Matrix) Matrix {
	r, c := getSize(m)
	out := make(Matrix, c)
	for j := 0; j < c; j++ {
		out[j] = make([]float64, r)
		for i := 0; i < r; i++ {
			out[j][i] = m[i][j]
		}
	}
	return out
}

// packMatrix returns a view (sub-slices) of a matrix block.
func packMatrix(matrix Matrix, a0, a1, b0, b1 int) Matrix {
	out := make(Matrix, b1-b0)
	for i := range out {
		out[i] = matrix[b0+i][a0:a1]
	}
	return out
}

const (
	defaultMB = 64
	defaultNB = 64
	defaultKB = 64
)

// gemmSequential is the original sequential implementation, kept for the fast path.
func gemmSequential(A, B Matrix, alpha float64, C Matrix, beta float64, MB, NB, KB int) (Matrix, error) {
	if err := validateMatrix(A, "A"); err != nil {
		return nil, err
	}
	if err := validateMatrix(B, "B"); err != nil {
		return nil, err
	}
	m, k := getSize(A)
	k2, n := getSize(B)
	if k != k2 {
		return nil, fmt.Errorf("shape mismatch: A is (%d,%d), B is (%d,%d)", m, k, k2, n)
	}
	if C != nil {
		if err := validateMatrix(C, "C"); err != nil {
			return nil, err
		}
		m2, n2 := getSize(C)
		if m2 != m || n2 != n {
			return nil, fmt.Errorf("C has shape (%d,%d); expected (%d,%d)", m2, n2, m, n)
		}
		if beta != 1.0 {
			for i := 0; i < m; i++ {
				for j := 0; j < n; j++ {
					C[i][j] *= beta
				}
			}
		}
	} else {
		C = generateMatrix(m, n)
	}
	if alpha == 0.0 {
		return C, nil
	}
	Bt := transpose(B)
	for n0 := 0; n0 < n; n0 += NB {
		n1 := n0 + NB
		if n1 > n {
			n1 = n
		}
		for k0 := 0; k0 < k; k0 += KB {
			k1 := k0 + KB
			if k1 > k {
				k1 = k
			}
			Bpack := packMatrix(Bt, k0, k1, n0, n1)
			for m0 := 0; m0 < m; m0 += MB {
				m1 := m0 + MB
				if m1 > m {
					m1 = m
				}
				Apack := packMatrix(A, k0, k1, m0, m1)
				kb := k1 - k0
				partialMatmul(Apack, Bpack, C, alpha, m0, n0, kb)
			}
		}
	}
	return C, nil
}

// Gemm computes C := alpha * A * B + beta * C using a parallel blocked algorithm.
func Gemm(A, B Matrix, alpha float64, C Matrix, beta float64, MB, NB, KB int) (Matrix, error) {
	if err := validateMatrix(A, "A"); err != nil {
		return nil, err
	}
	if err := validateMatrix(B, "B"); err != nil {
		return nil, err
	}
	m, k := getSize(A)
	k2, n := getSize(B)
	if k != k2 {
		return nil, fmt.Errorf("shape mismatch: A is (%d,%d), B is (%d,%d)", m, k, k2, n)
	}
	if C != nil {
		if err := validateMatrix(C, "C"); err != nil {
			return nil, err
		}
		m2, n2 := getSize(C)
		if m2 != m || n2 != n {
			return nil, fmt.Errorf("C has shape (%d,%d); expected (%d,%d)", m2, n2, m, n)
		}
	}

	if MB == 0 { MB = defaultMB }
	if NB == 0 { NB = defaultNB }
	if KB == 0 { KB = defaultKB }

	numMBlocks := (m + MB - 1) / MB
	numNBlocks := (n + NB - 1) / NB
	if numMBlocks*numNBlocks < runtime.NumCPU() {
		return gemmSequential(A, B, alpha, C, beta, MB, NB, KB)
	}

	if C == nil {
		C = generateMatrix(m, n)
	} else if beta != 1.0 {
		for i := 0; i < m; i++ {
			for j := 0; j < n; j++ {
				C[i][j] *= beta
			}
		}
	}

	if alpha == 0.0 {
		return C, nil
	}

	Bt := transpose(B)
	numWorkers := runtime.NumCPU()
	var wg sync.WaitGroup
	sem := make(chan struct{}, numWorkers)

	for m0 := 0; m0 < m; m0 += MB {
		m1 := m0 + MB
		if m1 > m {
			m1 = m
		}
		for n0 := 0; n0 < n; n0 += NB {
			n1 := n0 + NB
			if n1 > n {
				n1 = n
			}
			wg.Add(1)
			sem <- struct{}{}
			go func(m0, m1, n0, n1 int) {
				defer wg.Done()
				for k0 := 0; k0 < k; k0 += KB {
					k1 := k0 + KB
					if k1 > k {
						k1 = k
					}
					Bpack := packMatrix(Bt, k0, k1, n0, n1)
					Apack := packMatrix(A, k0, k1, m0, m1)
					kb := k1 - k0
					partialMatmul(Apack, Bpack, C, alpha, m0, n0, kb)
				}
				<-sem
			}(m0, m1, n0, n1)
		}
	}
	wg.Wait()
	return C, nil
}

File: llm_written/python_gemini/gemm/gemm_parallel.py

from typing import List, Tuple, Optional
import numbers
import concurrent.futures
import os

# --- Matrix Validation and Helpers (from original code) ---

def get_size(matrix: List[List[float]]) -> Tuple[int, int]:
    """Return (rows, cols) of a rectangular matrix."""
    return len(matrix), len(matrix[0])


def validate_matrix(matrix: List[List[float]], name: str = "Matrix"):
    """Validate matrix is a non-empty rectangular list-of-lists of numbers."""
    if not isinstance(matrix, list) or not matrix:
        raise ValueError(f"{name} must be a non-empty list of rows.")
    if not isinstance(matrix[0], list) or not matrix[0]:
        raise ValueError(f"{name} must have at least one column.")
    
    cols = len(matrix[0])
    for r, row in enumerate(matrix):
        if not isinstance(row, list):
            raise ValueError(f"{name} row {r} is not a list.")
        if len(row) != cols:
            raise ValueError(f"{name} is ragged: row {r} has {len(row)} cols, expected {cols}.")
        for c, v in enumerate(row):
            if not isinstance(v, numbers.Real):
                raise ValueError(f"{name}[{r}][{c}] is not a real number: {v!r}")


def generate_matrix(rows: int, cols: int) -> List[List[float]]:
    """Generates a matrix of zeros."""
    return [[0.0] * cols for _ in range(rows)]


def transpose(matrix: List[List[float]]) -> List[List[float]]:
    """Transposes a matrix."""
    return [list(col) for col in zip(*matrix)]


def pack_matrix(matrix: List[List[float]], col_start: int, col_end: int, row_start: int, row_end: int) -> List[List[float]]:
    """Extracts a submatrix (pack) for block processing."""
    return [row[col_start:col_end] for row in matrix[row_start:row_end]]


# --- Sequential Implementation ---

def _partial_matmul_sequential(
        A_pack: List[List[float]], 
        B_pack: List[List[float]], 
        C: List[List[float]],
        alpha: float,
        m0: int,
        n0: int,
        kb: int,
) -> None:
    """Helper for the sequential version to perform block multiplication."""
    for i_off, A_ik in enumerate(A_pack):
        C_i = C[m0 + i_off]
        for j_off, B_jk in enumerate(B_pack):
            s = 0.0
            for k in range(kb):
                s += A_ik[k] * B_jk[k]
            C_i[n0 + j_off] += alpha * s


def gemm_sequential(
        A: List[List[float]], 
        B: List[List[float]],
        alpha: float = 1.0, 
        C: Optional[List[List[float]]] = None, 
        beta: float = 0.0,
        MB: int = 64,
        NB: int = 64,
        KB: int = 64,
):
    """
    Sequential implementation of C := alpha * A * B + beta * C.
    This is the original provided algorithm, preserved for baseline and small inputs.
    """
    validate_matrix(A, "A")
    validate_matrix(B, "B")

    m, k = get_size(A)
    k2, n = get_size(B)

    if k != k2:
        raise ValueError(f"Shape mismatch: A is ({m},{k}), B is ({k2},{n}).")
    
    if C is not None:
        validate_matrix(C, "C")
        m2, n2 = get_size(C) 
        if m2 != m or n2 != n:
            raise ValueError(f"C has shape ({m2},{n2}); expected ({m},{n}).")
        if beta != 1.0:
            for i in range(m):
                for j in range(n):
                    C[i][j] *= beta  
    else:
        # If C is not provided, beta is assumed to be 0.
        C = generate_matrix(m, n)

    if alpha == 0.0:
        return C
    
    B_t = transpose(B)
    
    for n0 in range(0, n, NB):
        n1 = min(n0 + NB, n)
        for k0 in range(0, k, KB):
            k1 = min(k0 + KB, k)
            B_pack = pack_matrix(B_t, k0, k1, n0, n1)

            for m0 in range(0, m, MB):
                m1 = min(m0 + MB, m)
                A_pack = pack_matrix(A, k0, k1, m0, m1)

                kb = k1 - k0
                _partial_matmul_sequential(A_pack, B_pack, C, alpha, m0, n0, kb)
    
    return C


# --- Parallel Implementation ---

# Global variables for worker processes to hold shared matrices
_A_shared = None
_Bt_shared = None

def _init_worker(A: List[List[float]], B_t: List[List[float]]):
    """Initializer for each worker process in the pool."""
    global _A_shared, _Bt_shared
    _A_shared = A
    _Bt_shared = B_t

def _gemm_worker(args: Tuple) -> Tuple[int, int, List[List[float]]]:
    """
    Worker function that computes one block of the output matrix C.
    It reads from the global shared matrices _A_shared and _Bt_shared.
    """
    global _A_shared, _Bt_shared
    alpha, m0, n0, m1, n1, k, KB = args
    
    C_block_rows = m1 - m0
    C_block_cols = n1 - n0
    C_block = generate_matrix(C_block_rows, C_block_cols)

    for k0 in range(0, k, KB):
        k1 = min(k0 + KB, k)
        kb = k1 - k0

        A_pack = pack_matrix(_A_shared, k0, k1, m0, m1)
        B_pack = pack_matrix(_Bt_shared, k0, k1, n0, n1)

        for i_off, A_ik in enumerate(A_pack):
            C_i = C_block[i_off]
            for j_off, B_jk in enumerate(B_pack):
                s = 0.0
                for k_inner in range(kb):
                    s += A_ik[k_inner] * B_jk[k_inner]
                C_i[j_off] += s

    if alpha != 1.0:
        for r in range(C_block_rows):
            for c in range(C_block_cols):
                C_block[r][c] *= alpha

    return m0, n0, C_block


def gemm(
        A: List[List[float]], 
        B: List[List[float]],
        alpha: float = 1.0, 
        C: Optional[List[List[float]]] = None, 
        beta: float = 0.0,
        MB: int = 64,
        NB: int = 64,
        KB: int = 64,
):
    """
    Compute C := alpha * A * B + beta * C using a parallel block-based algorithm.

    A is (m x k), B is (k x n), C is (m x n). If C is None, a zero matrix is used.
    The operation is parallelized by dividing the output matrix C into blocks
    and computing each block in a separate process.
    """
    validate_matrix(A, "A")
    validate_matrix(B, "B")

    m, k = get_size(A)
    k2, n = get_size(B)

    if k != k2:
        raise ValueError(f"Shape mismatch: A is ({m},{k}), B is ({k2},{n}).")

    # Fast path for small matrices where parallel overhead is not worth it.
    num_m_blocks = (m + MB - 1) // MB
    num_n_blocks = (n + NB - 1) // NB
    # Use sequential version if the number of tasks is less than or equal to the number of available cores.
    if num_m_blocks * num_n_blocks <= (os.cpu_count() or 1):
        return gemm_sequential(A, B, alpha, C, beta, MB, NB, KB)

    # Handle the output matrix C and the beta scaling factor.
    if C is not None:
        validate_matrix(C, "C")
        m2, n2 = get_size(C) 
        if m2 != m or n2 != n:
            raise ValueError(f"C has shape ({m2},{n2}); expected ({m},{n}).")
        if beta != 1.0:
            for i in range(m):
                for j in range(n):
                    C[i][j] *= beta  
    else:
        # If C is not provided, beta is assumed to be 0.
        C = generate_matrix(m, n)

    if alpha == 0.0:
        return C
    
    B_t = transpose(B)
    
    tasks = []
    for n0 in range(0, n, NB):
        n1 = min(n0 + NB, n)
        for m0 in range(0, m, MB):
            m1 = min(m0 + MB, m)
            tasks.append((alpha, m0, n0, m1, n1, k, KB))

    num_workers = os.cpu_count() or 1
    with concurrent.futures.ProcessPoolExecutor(
        max_workers=num_workers,
        initializer=_init_worker,
        initargs=(A, B_t)
    ) as executor:
        
        # The results can be processed as they are completed.
        # The order of addition to C does not matter for correctness.
        futures = [executor.submit(_gemm_worker, task) for task in tasks]
        for future in concurrent.futures.as_completed(futures):
            m0_res, n0_res, C_block = future.result()
            
            if not C_block or not C_block[0]:
                continue

            m1_res = m0_res + len(C_block)
            n1_res = n0_res + len(C_block[0])

            # Add the computed block to the final matrix C
            for i in range(m0_res, m1_res):
                for j in range(n0_res, n1_res):
                    C[i][j] += C_block[i - m0_res][j - n0_res]
    
    return C

if __name__ == '__main__':
    # Example usage guard
    A_sample = [[1.0, 2.0, 3.0], [4.0, 5.0, 6.0], [7.0, 8.0, 9.0]]
    B_sample = [[1.0, 4.0], [2.0, 5.0], [3.0, 6.0]]
    
    print("Running sample matrix multiplication...")
    C_result = gemm(A_sample, B_sample, alpha=1.0)
    
    print("Matrix A:")
    for row in A_sample:
        print(row)
    print("\nMatrix B:")
    for row in B_sample:
        print(row)
    print("\nResult C = A * B:")
    for row in C_result:
        print([round(x, 1) for x in row])

    # Expected: [[14.0, 32.0], [32.0, 77.0], [50.0, 122.0]]
